File: bms/go/geo/geo_s2_CellID_PrevWrap/geo_s2_CellID_PrevWrap.go
package test

const (
	geo_s2_NumFaces   = 6
	geo_s2_MaxLevel   = 30
	geo_s2_PosBits    = 2*geo_s2_MaxLevel + 1
	geo_s2_wrapOffset = uint64(geo_s2_NumFaces) << geo_s2_PosBits
)

func (ci geo_s2_CellID) PrevWrap() geo_s2_CellID {
	p := ci.Prev()
	if uint64(p) < geo_s2_wrapOffset {
		return p
	}
	return geo_s2_CellID(uint64(p) + geo_s2_wrapOffset)
}

func (ci geo_s2_CellID) Prev() geo_s2_CellID {
	return geo_s2_CellID(uint64(ci) - ci.lsb()<<1)
}

func (ci geo_s2_CellID) lsb() uint64 { return uint64(ci) & -uint64(ci) }

type geo_s2_CellID uint64

File: translations_only/fail/go/geo/geo_s2_CellID_PrevWrap/claude3/CAPR/geo_s2_CellID_PrevWrap.rs

const GEO_S2_NUM_FACES: u64 = 6;
const GEO_S2_MAX_LEVEL: u64 = 30;
const GEO_S2_POS_BITS: u64 = 2 * GEO_S2_MAX_LEVEL + 1;
const GEO_S2_WRAP_OFFSET: u64 = GEO_S2_NUM_FACES << GEO_S2_POS_BITS;

type GeoS2CellID = u64;

fn prev_wrap(ci: GeoS2CellID) -> GeoS2CellID {
    let p = prev(ci);
    if p < GEO_S2_WRAP_OFFSET {
        return p;
    }
    return p + GEO_S2_WRAP_OFFSET;
}

fn prev(ci: GeoS2CellID) -> GeoS2CellID {
    return ci - lsb(ci) << 1;
}

fn lsb(ci: GeoS2CellID) -> GeoS2CellID {
    return ci & (!ci + 1);
}
